airflow/contrib/hooks/gcp_sql_hook.py

```diff
@@ -261,3 +278,541 @@ def _wait_for_operation_to_complete(self, project_id, operation_name):
+        """
+        super(CloudSqlProxyRunner, self).__init__()
+        self.path_prefix = path_prefix
+        if not self.path_prefix:
+            raise AirflowException("The path_prefix must not be empty!")
+        self.sql_proxy_was_downloaded = False
+        self.sql_proxy_version = sql_proxy_version
+        self.download_sql_proxy_dir = None
+        self.sql_proxy_process = None
+        self.instance_specification = instance_specification
+        self.project_id = project_id
+        self.gcp_conn_id = gcp_conn_id
+        self.command_line_parameters = []
+        self.cloud_sql_proxy_socket_directory = self.path_prefix
+        self.sql_proxy_path = sql_proxy_binary_path if sql_proxy_binary_path \
+            else self.path_prefix + "_cloud_sql_proxy"
+        self.credentials_path = self.path_prefix + "_credentials.json"
+        self._build_command_line_parameters()
+
+    def _build_command_line_parameters(self):
+        self.command_line_parameters.extend(
+            ['-dir', self.cloud_sql_proxy_socket_directory])
+        self.command_line_parameters.extend(
+            ['-instances', self.instance_specification])
+
+    @staticmethod
+    def _is_os_64bit():
+        return platform.machine().endswith('64')
+
+    def _download_sql_proxy_if_needed(self):
+        if os.path.isfile(self.sql_proxy_path):
+            self.log.info("cloud-sql-proxy is already present")
+            return
+        system = platform.system().lower()
+        processor = "amd64" if CloudSqlProxyRunner._is_os_64bit() else "386"
+        if not self.sql_proxy_version:
+            download_url = CLOUD_SQL_PROXY_DOWNLOAD_URL.format(system, processor)
+        else:
+            download_url = CLOUD_SQL_PROXY_VERSION_DOWNLOAD_URL.format(
+                self.sql_proxy_version, system, processor)
+        proxy_path_tmp = self.sql_proxy_path + ".tmp"
+        self.log.info("Downloading cloud_sql_proxy from {} to {}".
+                      format(download_url, proxy_path_tmp))
+        r = requests.get(download_url, allow_redirects=True)
+        with open(proxy_path_tmp, 'wb') as f:
+            f.write(r.content)
+        if r.status_code != 200:
+            raise AirflowException(
+                "The cloud-sql-proxy could not be downloaded. Status code = {}. "
+                "Reason = {}".format(r.status_code, r.reason))
+        self.log.info("Moving sql_proxy binary from {} to {}".format(
+            proxy_path_tmp, self.sql_proxy_path
+        ))
+        shutil.move(proxy_path_tmp, self.sql_proxy_path)
+        os.chmod(self.sql_proxy_path, 0o744)  # Set executable bit
+        self.sql_proxy_was_downloaded = True
+
+    @provide_session
+    def _get_credential_parameters(self, session):
+        connection = session.query(models.Connection). \
+            filter(models.Connection.conn_id == self.gcp_conn_id).first()
+        session.expunge_all()
+        if GCP_CREDENTIALS_KEY_PATH in connection.extra_dejson:
```

Comment:
I looked at this before and I decided against. The hook's _get_credentials method returns google.oauth2.service_account.Credentials object which is not much use in our case - we have to really get the data either from loaded dictionary object or from a json file. We have no other way to pass credentials to cloud_sql_proxy but passing it credentials .json file, so we actually have to always either use existing or generate the file with credentials from whatever is stored in the database. We cannot do anything with Credentials object (well we could dump it back to json but I think that's an overkill to convert it from .json string/file to Credentials and then back to .json file. 

Another option would be to have a separate GCP credentials helper used by both gcp_api_base_hook and the gcp_sql_hook. Such helper could perform conversion from any type of stored credentials to any other (Credentials/Connection/Dictionary/JSON file). But I think that might be a separate pull request - it would require to touch GPC_API hook implementation which is used in other operators and likely requires more thorough testing. We could maybe add it to the backlog of "improvements" we want to do if we have time left?

Reply:
Sure, let's add it to the backlog.